File: reproto/src/reproto/show.py
# SPDX-FileCopyrightText: Frederic Ruget <fred@atlant.is> (GitHub: @douzebis)
#
# SPDX-License-Identifier: MIT

from pathlib import Path, PurePosixPath
from typing import cast

import yaml
from pyvis.network import Network


def _fqdn_matches_any(fqdn: str, patterns: tuple[str, ...]) -> bool:
    for p in patterns:
        if PurePosixPath(f'/{fqdn}').full_match(f'/{p}'):
            return True
    return False


def _make_network(title: str) -> Network:
    """Create a pyvis Network with shared settings."""
    net = Network(
        height="90vh",
        width="100%",
        directed=True,
        bgcolor="#222222",
        cdn_resources='in_line',
    )
    net.heading = ''
    net.barnes_hut()
    net.set_options("""{
  "nodes": {
    "font": { "face": "sans-serif", "color": "#88cc88", "size": 13 }
  },
  "edges": {
    "font": { "face": "sans-serif", "color": "#888888", "size": 11,
              "align": "middle" }
  },
  "physics": {
    "enabled": true,
    "barnesHut": {
      "gravitationalConstant": -8000,
      "centralGravity": 0.3,
      "springLength": 100,
      "springConstant": 0.04,
      "damping": 0.09,
      "avoidOverlap": 1
    },
    "minVelocity": 0.75
  }
}""")
    return net


# 5-step Fibonacci brightness scale for top-level (named) nodes.
# Each step covers a Fibonacci range of merged-type counts:
#   step 1: count == 1          (single)
#   step 2: count == 2
#   step 3: count in 3..4
#   step 4: count in 5..7
#   step 5: count >= 8
# Colours are amber/gold, evenly spread so adjacent steps are clearly distinct.
_BRIGHTNESS_STEPS = [
    (1, '#664400'),   # step 1 — darkest amber
    (2, '#995500'),   # step 2
    (3, '#cc8800'),   # step 3
    (5, '#ffaa00'),   # step 4
    (8, '#ffdd44'),   # step 5 — brightest gold
]


def _node_colour(fqdns: list[str]) -> str:
    """Compute node colour for a non-leaf message node.

    Top-level nodes (have at least one FQDN) use an amber/gold hue on a
    5-step Fibonacci brightness scale keyed on merged-type count.
    Internal nodes (no FQDN) use dim blue.
    """
    count = len(fqdns)
    if count == 0:
        return '#3355aa'  # dim blue — internal node with no named root
    colour = _BRIGHTNESS_STEPS[0][1]
    for threshold, c in _BRIGHTNESS_STEPS:
        if count >= threshold:
            colour = c
    return colour


def render_scoring_graph(
    compiled_yaml: str,
    output_path: Path,
    title: str,
    with_leaf_nodes: bool = False,
    hide: tuple[str, ...] = (),
) -> None:
    """Render a compiled scoring graph (states/transitions/roots YAML) to HTML."""
    data = cast(dict, yaml.safe_load(compiled_yaml))
    net = _make_network(title)

    # Determine which states have outgoing transitions (non-leaf) vs leaf.
    states_with_outgoing: set[int] = {t['from'] for t in data.get('transitions', [])}
    leaf_states: set[int] = {
        s['id'] for s in data.get('states', []) if s['id'] not in states_with_outgoing
    }

    # Build tooltip: state_id -> sorted list of FQDNs from roots.
    state_fqdns: dict[int, list[str]] = {}
    for root in data.get('roots', []):
        sid = root['state']
        state_fqdns.setdefault(sid, []).append(root['fqdn'])
    for fqdns in state_fqdns.values():
        fqdns.sort()

    # Compute hidden state IDs from --hide patterns.
    hidden_ids: set[int] = set()
    if hide:
        for sid, fqdns in state_fqdns.items():
            if any(_fqdn_matches_any(f, hide) for f in fqdns):
                hidden_ids.add(sid)

    for state in data.get('states', []):
        sid = state['id']
        wt = state['wire_type']
        is_string = state.get('is_string', False)
        is_leaf = sid in leaf_states

        if sid in hidden_ids:
            continue
        if is_leaf and not with_leaf_nodes:
            continue

        fqdns = state_fqdns.get(sid, [])
        tooltip = '\n'.join(fqdns) if fqdns else ''
        # Label: short type name for named nodes, "+N" suffix when merged, blank for internal.
        if len(fqdns) == 1:
            node_label = fqdns[0].rsplit('.', 1)[-1]
        elif len(fqdns) > 1:
            node_label = f'{fqdns[0].rsplit(".", 1)[-1]}+{len(fqdns) - 1}'
        else:
            node_label = ''

        if not is_leaf:
            net.add_node(
                sid,
                label=node_label,
                title=tooltip,
                shape='dot',
                color=_node_colour(fqdns),
                size=20,
            )
        else:
            if wt == 2 and is_string:
                label = 'string'
                colour = '#ff8844'
            elif wt == 0:
                label = 'varint'
                colour = '#ffcc44'
            elif wt == 1:
                label = 'i64'
                colour = '#ffcc44'
            elif wt == 2:
                label = 'len'
                colour = '#ffcc44'
            elif wt == 5:
                label = 'i32'
                colour = '#ffcc44'
            else:
                label = str(wt)
                colour = '#ffcc44'
            net.add_node(
                sid,
                label=label,
                title=tooltip,
                shape='square',
                color=colour,
                size=12,
            )

    edge_colours = {
        'optional': '#4444ff',
        'repeated': '#44aaff',
        'packed':   '#884488',
        'required': '#cc2222',
    }
    for t in data.get('transitions', []):
        if t['from'] in hidden_ids or t['to'] in hidden_ids:
            continue
        if not with_leaf_nodes and t['to'] in leaf_states:
            continue
        colour = edge_colours.get(t['label'], '#888888')
        net.add_edge(
            t['from'],
            t['to'],
            label=str(t['field']),
            color=colour,
        )

    net.write_html(str(output_path))
    # Patch tooltip CSS: pyvis uses white-space:nowrap which prevents line
    # breaks in multi-FQDN tooltips.  Switch to pre-wrap so \n is respected.
    html = output_path.read_text()
    html = html.replace('white-space:nowrap', 'white-space:pre-wrap', 1)
    output_path.write_text(html)

File: score-graph/src/build_scoring_graph/show.rs
// SPDX-FileCopyrightText: Frederic Ruget <fred@atlant.is> (GitHub: @douzebis)
//
// SPDX-License-Identifier: MIT

//! Interactive HTML visualisation of the compiled scoring graph (vis-network).

use std::fmt::Write as FmtWrite;
use std::path::Path;

use super::graph::RawGraph;
use super::hopcroft::Partition;
use super::serial::CompiledGraph;

// ── Node label/title helpers ──────────────────────────────────────────────────

/// Returns `(label, title)` for a leaf node given its wire_type and attributes.
fn leaf_label_title(
    wire_type: u8,
    is_string: bool,
    enum_range_idx: u16,
    enum_ranges: &[(i32, i32)],
) -> (String, String) {
    if enum_range_idx != 0xFFFF {
        if let Some(&(min, max)) = enum_ranges.get(enum_range_idx as usize) {
            return ("ENUM".into(), format!("{min}..{max}"));
        }
        return ("ENUM?".into(), String::new());
    }
    match wire_type {
        0 => ("VARINT".into(), String::new()),
        1 => ("FIXED64".into(), String::new()),
        2 if is_string => ("STRING".into(), String::new()),
        2 => ("LENDEL".into(), String::new()),
        3 => ("GROUP".into(), String::new()),
        5 => ("FIXED32".into(), String::new()),
        _ => ("?".into(), String::new()),
    }
}

// ── ScoringKind colour palette ────────────────────────────────────────────────

fn wire_type_color(wire_type: u8) -> &'static str {
    match wire_type {
        0 => "#888888", // VARINT / ENUM
        1 => "#884400", // I64
        2 => "#0066cc", // LEN_*
        3 => "#cc6600", // GROUP
        5 => "#446600", // I32
        _ => "#ffffff",
    }
}

// ── Entry ─────────────────────────────────────────────────────────────────────

pub fn write_html(
    path: &Path,
    raw: &RawGraph,
    partition: &Partition,
    compiled: &CompiledGraph,
) -> Result<(), Box<dyn std::error::Error>> {
    let html = build_html(raw, partition, compiled);
    std::fs::write(path, html)?;
    Ok(())
}

const SHOW_LEAVES: bool = false;

fn build_html(raw: &RawGraph, partition: &Partition, compiled: &CompiledGraph) -> String {
    // Determine which state IDs are leaf nodes (no outgoing transitions).
    let non_leaf_states: std::collections::HashSet<u32> =
        compiled.transitions.iter().map(|t| t.state_id).collect();
    let num_msg_blocks = compiled
        .nodes
        .iter()
        .filter(|n| non_leaf_states.contains(&n.state_id))
        .count();

    // ── Build nodes JSON ──────────────────────────────────────────────────────
    // One node per state (post-Hopcroft), plus leaf nodes, plus one entry
    // node per root FQDN (with IDs starting beyond the state/leaf range).
    let mut state_labels: Vec<Vec<String>> = vec![Vec::new(); partition.num_blocks()];
    for (fqdn, &node_id) in &raw.node_ids {
        let block = partition.block_of(node_id) as usize;
        state_labels[block].push(fqdn.clone());
    }

    let mut nodes_js = String::new();
    // Message states — plain blue dots
    for (block, block_labels) in state_labels[..num_msg_blocks].iter().enumerate() {
        let mut labels = block_labels.clone();
        labels.sort();
        let short = labels
            .iter()
            .min_by_key(|s| s.len())
            .and_then(|s| s.rsplit('.').next())
            .unwrap_or("?");
        let label = if labels.len() > 1 {
            format!("{short}+")
        } else {
            short.to_string()
        };
        let title = {
            let mut sorted = labels.clone();
            sorted.sort();
            format!("{block}<br>{}", sorted.join("<br>"))
        };
        let _ = writeln!(
            nodes_js,
            "  {{id:{block}, label:{label:?}, title:{title:?}, \
             color:'#aaaaff', shape:'dot', size:15}},"
        );
    }
    // Leaf nodes — derived from compiled.nodes where no outgoing transitions exist
    if SHOW_LEAVES {
        for node in compiled.nodes.iter() {
            let id = node.state_id as usize;
            if non_leaf_states.contains(&(id as u32)) {
                continue;
            }
            let (label, title) = leaf_label_title(
                node.wire_type,
                node.is_string,
                node.enum_range_idx,
                &compiled.enum_ranges,
            );
            let _ = writeln!(
                nodes_js,
                "  {{id:{id}, label:{label:?}, title:{title:?}, \
                 color:'#ffcc44', shape:'square', size:10}},"
            );
        }
    }
    // Entry nodes — one diamond per root FQDN
    let entry_id_base = partition.num_blocks(); // IDs beyond state/leaf range
    for (i, root) in compiled.roots.iter().enumerate() {
        let id = entry_id_base + i;
        let short = root.fqdn.rsplit('.').next().unwrap_or(&root.fqdn);
        let _ = writeln!(
            nodes_js,
            "  {{id:{id}, label:{short:?}, title:{fqdn:?}, \
             color:'#ff9944', shape:'diamond', size:20}},",
            fqdn = root.fqdn,
        );
    }

    // ── Build edges JSON ──────────────────────────────────────────────────────
    let mut edges_js = String::new();
    // Transition edges
    for t in &compiled.transitions {
        let child = t.child_state_id;
        if !SHOW_LEAVES && !non_leaf_states.contains(&child) {
            continue;
        }
        let field_num = t.field_number;
        // Derive edge color from child node's wire_type.
        let child_wt = compiled
            .nodes
            .iter()
            .find(|n| n.state_id == child)
            .map_or(0u8, |n| n.wire_type);
        let color = wire_type_color(child_wt);
        let dst = child;
        let _ = writeln!(
            edges_js,
            "  {{from:{src}, to:{dst}, label:'f{field_num}', \
             color:{{color:{color:?}}}, arrows:'to', font:{{size:10}}}},",
            src = t.state_id,
        );
    }
    // Entry → state edges (dashed, no label)
    for (i, root) in compiled.roots.iter().enumerate() {
        let from = entry_id_base + i;
        let to = root.state_id;
        let _ = writeln!(
            edges_js,
            "  {{from:{from}, to:{to}, label:'', \
             color:{{color:'#ff9944'}}, arrows:'to', dashes:true}},",
        );
    }

    // ── Build legend entries ──────────────────────────────────────────────────
    let stats = format!(
        "States: {} &nbsp;|&nbsp; Transitions: {} &nbsp;|&nbsp; Roots: {}",
        compiled.num_states,
        compiled.transitions.len(),
        compiled.roots.len(),
    );

    // ── Assemble HTML ─────────────────────────────────────────────────────────
    format!(
        r#"<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<title>Scoring Graph</title>
<script src="https://cdnjs.cloudflare.com/ajax/libs/vis/4.21.0/vis.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/vis/4.21.0/vis.min.css">
<style>
  body {{ margin:0; background:#222; color:#eee; font-family:sans-serif; }}
  #graph {{ width:100%; height:92vh; border:none; }}
  #bar {{ padding:6px 12px; background:#333; font-size:13px; }}
  #legend {{ display:inline-block; margin-left:20px; }}
  .dot {{ display:inline-block; width:12px; height:12px; border-radius:50%;
           margin-right:4px; vertical-align:middle; }}
  .sq  {{ display:inline-block; width:12px; height:12px;
           margin-right:4px; vertical-align:middle; }}
  .dia {{ display:inline-block; width:12px; height:12px;
           transform:rotate(45deg); margin-right:6px; vertical-align:middle; }}
</style>
</head>
<body>
<div id="bar">
  {stats}
  &nbsp;&nbsp;
  <span id="legend">
    <span class="dia" style="background:#ff9944"></span>Entry point&nbsp;&nbsp;
    <span class="dot" style="background:#aaaaff"></span>Message state&nbsp;&nbsp;
    <span class="sq"  style="background:#ffcc44"></span>Leaf (scalar)&nbsp;&nbsp;
    <span style="color:#888888">&#9644;</span> VARINT &nbsp;
    <span style="color:#884400">&#9644;</span> I64 &nbsp;
    <span style="color:#0066cc">&#9644;</span> LEN &nbsp;
    <span style="color:#cc6600">&#9644;</span> GROUP &nbsp;
    <span style="color:#446600">&#9644;</span> I32
  </span>
</div>
<div id="graph"></div>
<script>
var nodes = new vis.DataSet([
{nodes_js}]);
var edges = new vis.DataSet([
{edges_js}]);
var container = document.getElementById('graph');
var data = {{ nodes: nodes, edges: edges }};
var options = {{
  physics: {{
    enabled: true,
    barnesHut: {{
      gravitationalConstant: -8000,
      centralGravity: 0.3,
      springLength: 120,
      springConstant: 0.04,
      damping: 0.09,
      avoidOverlap: 1
    }},
    minVelocity: 0.75,
    stabilization: {{ enabled: false }}
  }},
  nodes: {{ font: {{ color: '#eee' }} }}
}};
var network = new vis.Network(container, data, options);
network.on('stabilizationIterationsDone', function() {{
  network.setOptions({{ physics: {{ enabled: false }} }});
}});
</script>
</body>
</html>
"#,
    )
}
